### ISSUE TITLE
Release 2.0

### ISSUE THREAD

**Issue by fbourigault** [symfony-cmf/symfony-cmf#235]
I'm quite new at Symfony CMF contributing, but I think Symfony CMF lack a stable release compatible with Symfony 3 which was released a year ago. I don't know what work is required to get a stable 2.0 release, but I think we need a plan.

This plan should, IMHO, focus on breaking things. New features could be delayed in the 2.X series. For example, we started to extract FormTypes for sonata Admin classes. In hindsight, this task is too big and we should focus on moving sonata stuff in it's own bundle, and then as new feature adding form types.

Does this could be applied to others things wanted for 2.0 ?

WDYT?

**Comment by ElectricMaxxx** [comment]
> This plan should, IMHO, focus on breaking things. New features could be delayed in the 2.X series. For example, we started to extract FormTypes for sonata Admin classes. In hindsight, this task is too big and we should focus on moving sonata stuff in it's own bundle, and then as new feature adding form types.

:+1: right, but we need to remove the sonata stuff to do the break right now and maybe have a sonata-integration-bundle on 2.1

> Does this could be applied to others things wanted for 2.0 ?

Major thing should be a new release with the Symfony 3.X support.

**Comment by dbu** [comment]
hi, thanks for your input. i am rather unhappy by how much we are dragging that release out. the master branches mostly work with symfony 3 already. we "just" need to wrap up things. there is a rough plan at https://github.com/symfony-cmf/symfony-cmf/wiki/Release-2.0 but we don't have an overview of the state of each bundle. i will try to check on the important bundles that we have the milestone properly set up and link to those milestones from the wiki page.
the main effort i am aware of is taking out the sonata admins to an own bundle. you started work on that a while ago in https://github.com/symfony-cmf/routing-bundle/pull/361. the other bundles need a similar treatment (which hopefully becomes easier once we solved the questions for one of the bundles)

**Comment by fbourigault** [comment]
> the main effort i am aware of is taking out the sonata admins to an own bundle. you started work on that a while ago in symfony-cmf/routing-bundle#361. the other bundles need a similar treatment (which hopefully becomes easier once we solved the questions for one of the bundles)

I didn't yet moved any sonata admins to an own bundle. I just extracted FormTypes out those Admins. As this looks to be the main effort to provide, I suggest to move Admin classes in the sonata integration one without extracting FormTypes and then write those FormTypes as a new feature for 2.X releases.

**Comment by dbu** [comment]
how much is left to wrap up https://github.com/symfony-cmf/routing-bundle/pull/361? if its too much effort then indeed, we should just move stuff unchanged and release and move on from that. its no BC break on the main bundles if we add form types later - might be a BC break for the sonata admins integration bundle though.

**Comment by fbourigault** [comment]
I think [this comment](https://github.com/symfony-cmf/routing-bundle/pull/361#issuecomment-250910768) from @wouterj is a good resume of how difficult the use of extracted FormTypes into the sonata classes is difficult.

The main issue is that is we create a FormType, the sonata only steal it fields and does not use the main FormType and so any behavior (events) is lost. As a resume, we introducted nice FormTypes but sonata can't use it at their full power. This worry me and should be solved by writing a new FormMapper to allow use of a FormType.

**Comment by ElectricMaxxx** [comment]
Just moving the sonata stuff out to the integration-bundle isn't that difficult - just some copy and paste. I did it on seo-bundle and on routing-bundle, before i started to create the form type stuff. I had it working in some hours. So I think it should be possible to simply extract (not just remove) the sonata stuff in one hackday or similar. Adding the form type stuff can be done afterwards.

**Comment by wouterj** [comment]
What about this:

 * Forget extracting for the moment, just moving the complete admin classes to the SonataAdminIntegrationBundle
 * Deprecate the admin classes in the main bundles

Then, we can release 2.0. Adding form types means new features, so this can be done in 2.x. Removing the deprecated admins in the main bundle can be done in 3.x when we've finished extracting.

**Comment by fbourigault** [comment]
About deprecating admin classes in the main bundles, should we empty the class and make it extends the one provided by the integration bundle?
Does, the integration bundle will be a dependency of any main bundles and then just be a suggested bundle in the 3.X series?

**Comment by wouterj** [comment]
Yeah, that's a quite complex one. I would rather not introduce the dependency, as we can't checker whether people have required it explicitly (and thus, we can't garantee that no deprecation means upgrading will be without a problem). On the other hand, maintaining 2 classes seems undoable...

**Comment by fbourigault** [comment]
Could we remove the dependency from the main bundle with 2.0 and do the breaking for 2.0 by requiring users to use the sonata integration bundle to continue using admins?

**Comment by dbu** [comment]
>   * Forget extracting for the moment, just moving the complete admin
>     classes to the SonataAdminIntegrationBundle

+1

>   * Deprecate the admin classes in the main bundles

what is the point of keeping the admins in the main bundles? if we move
them, we can simply delete them and people that upgrade and want sonata
can use the SonataAdminIntegrationBundle. that bundle is 1.0 and once we
made form types for all admin things in the bundles, we do a version 2
that uses those.

**Comment by wouterj** [comment]
Alright, that's possible as well. 👍

**Comment by wouterj** [comment]
I've created a spreadsheat with all bundles and the things that have to be done before the release: https://docs.google.com/spreadsheets/d/1taPM0QrKhRsbsTR9LgLW-cFzKj41x4XbhuxfWKXfz9M/edit?usp=sharing

Let's all work on the missing bits the coming 2 weeks and try to get a 2.0 release out this year. Wdyt?

**Comment by dbu** [comment]
great, thanks for that spreadsheet. yeah, lets make an effort and get
the symfony community an early christmas present ;-)

i will be travelling on thursday, will make sure to mark what i am
working on. i propose we create git issues for the tasks and put the
number into the spreadsheet and comment on them / assign to ourselves
before starting on a task, to avoid duplicating the work. too many open
things on the same repository are probably also a bad idea.

**Comment by wouterj** [comment]
I started to add my name (`@wouterj`) to the spreadsheet to mark things I'm working on. Please also note that column H to the end cannot be executed before B-G are done. So let's focus on those first.

**Comment by ElectricMaxxx** [comment]
I think we also can use our waffle.io Board instead of the spreadsheet?
We simply need to create issues in each package and put the right label
on it. I think we would have a better overview then.

But before starting a discussion about the tool i will start working
too. Will pick little tasks now and invest some more time on our
slackday on friday.


greetz max

**Comment by ElectricMaxxx** [comment]
I think this should be the right board: https://waffle.io/symfony-cmf/symfony-cmf?milestone=2.0

**Comment by ElectricMaxxx** [comment]
btw: working with issues and milestones out of one package for all of them (would be the half way into a monolithic repo) would be easier to visualize on a generalized board.

**Comment by ElectricMaxxx** [comment]
So i did that:
- created/updates milesone 2.0 due date on the 24. of december
- created/updated milestone 2.1
- moved all issues/PR to 2.1 which are not in the sheet
- created all missing issues and tagged with 2.0
for all relevant packages in the symfony-cmf repo. I have no right on phpcr and sonata to remove the 2.0 milestones there. When that is done, the board mentioned above will give a nice overview of our current work.

I just need one hint of  @wouterj  what he means with `Upgrade Files`

**Comment by dbu** [comment]
regarding the create-bundle in the google spreadsheet: did we not decide to abandon createbundle to save resources?

**Comment by ElectricMaxxx** [comment]
One suggestion: remove sonata first, then most of the deprecation warnings a gone :-) maybe it is on sonata-admin-integration, but we can fix them at once.

**Comment by ElectricMaxxx** [comment]
Shouldn't we add support until 3.2 now?

**Comment by ElectricMaxxx** [comment]
Just wanted to report my current state:

i merged the sonata extraction in seo-bundle. It is completely included in sonata-integration-bundle. For the other bundles i would recoment to clear all single bundles before moving the code to the sonata-admin-integration. Especially pushing that bundle to phpcr-odm 2.0 is an extreme dependency hell.

With clearing the bundles i mean all tasts of the spreadshirt including the update of phpcr-odm. I hope we can fix the sonata-admin-integration dependency hell then.

**Comment by wouterj** [comment]
I saw a PR updating the file structure, let's do that at the end, just before the release, to avoid merge conflicts and such.

> I just need one hint of @wouterj what he means with Upgrade Files

Creating UPGRADE files, like https://github.com/symfony-cmf/menu-bundle/blob/master/UPGRADE-2.0.md , to ease upgrading from 1.3 to 2.0.

Regarding your last comment, huge PRs are a nightmare. Let's do things bundle per bundle. If we have failing tests on the integration bundle because of dependencies, I suggest ignoring them for now and fixing them at the end.

**Comment by ElectricMaxxx** [comment]
> Regarding your last comment, huge PRs are a nightmare.
Indeed mixing the sonata extraction and the creation of the form type wasn't very good.

> Creating UPGRADE file
Found that by comparing ready with not ready repositories and added https://github.com/symfony-cmf/seo-bundle/pull/312/commits/b52f618e986461e36c450a53454f8dc7478b2e2a and https://github.com/symfony-cmf/seo-bundle/pull/312/commits/d0f372219015368ed46b09eec410b58add4a6b34 Should we inform that deep in each repository?

What about the Symfony Version? We have 2.8 = 3.0 and some 3.x in our Roadmap. I really would like serve a 3.2 support, cause it took us that long to serve 2.0, what do you think? I tried to visualize the validation-bug in https://github.com/symfony-cmf/routing-bundle/pull/372 and found some little issues to fix, isn't it?

**Comment by wouterj** [comment]
Well, 3.x should be compatible with 3.0. And 3.0 is compatible with 2.8. So to me, Symfony support is very easy for 2.0: Symfony 2.8, 3.0, 3.1, 3.2, 3.3 and 3.4 should be supported.

**Comment by ElectricMaxxx** [comment]
Right then let's put it explicit into the sheet or my board. Cause there will be some improvements to get it working. Did an example for routing bundle to see explicit symfony versions.

**Comment by ElectricMaxxx** [comment]
I also faced 
https://github.com/sonata-project/SonataAdminBundle/issues/4152
several times. Even when the usage of that `renderer` property seems to wiered, i think the change at
https://github.com/symfony/symfony/commit/b515702fc90d759fba502b7e474f7462b4a9b578
is a kind of a BC-Break, The property hasn't been private before as it wasn't defined.

**Comment by fbourigault** [comment]
It will be fixed in 3.2.1. https://github.com/symfony/symfony/pull/20769

**Comment by wouterj** [comment]
And SonataAdminBundle is also fixing it in a next release.

**Comment by ElectricMaxxx** [comment]
I started converting into short array notation only and found: https://github.com/a-vasyliev/php-short-array-syntax-converter. Two little commands:
```
maximilian@maximilian-ThinkPad-T540p:~/OpenSource/php-short-array-syntax-converter$ find ../Cmf/seo-bundle/src/ -name "*.php" -exec bin/array-converter -w "{}" \;
maximilian@maximilian-ThinkPad-T540p:~/OpenSource/php-short-array-syntax-converter$ find ../Cmf/seo-bundle/tests/ -name "*.php" -exec bin/array-converter -w "{}" \;
```
and it was done.

**Comment by ElectricMaxxx** [comment]
forget my last comment. We can't replace old array notation until we stop 1.* support in May 2018.

**Comment by ShinDarth** [comment]
is there anything simple that a beginner can use to start contributing to this project?

**Comment by dbu** [comment]
hi francesco and welcome!

> is there anything simple that a beginner can use to start contributing
> to this project?

good question. as you see in this issue, we are in the midst of wrapping
up the 2.0 versions. while most of those tasks are not hard, they are
also not very interesting and max, wouter and i know exactly what to do.
i would say that if you are using the cmf already, the most important
and doable contribution at this point would be trying to upgrade to
version 2. we started adding UPGRADE.md files to the repositories but
they are probably not complete yet. if you try to upgrade and point out
where things are missing in the upgrade file, or if something is not
understandable, that would help a lot!

**Comment by ShinDarth** [comment]
Thanks. Unfortunately I'm not using this already, but I can give a try

**Comment by ElectricMaxxx** [comment]
Yea. I think we have moved all sonata related stuff, or did I miss something?there are some issues open regarding configuration, but `sonata-admin-integration-bundle` is on its way. 

The other bundles/components mostly need its new file structure and the entries in the `UPGRADE.md`.  We should also have a look to align the travis files. Doing so we will have Symfony 3.2 support on the fly. 

The sonata stuff (and others) needs its implementation in the sandbox and then the last big stone needs to be moved: the documentation. Has anybody an overview what is to do there?

**Comment by ElectricMaxxx** [comment]
@ShinDarth what about: https://github.com/symfony-cmf/sonata-admin-integration-bundle/issues/42

**Comment by ElectricMaxxx** [comment]
@dbu i have no "settings" option on sonata-admin-integration-bundle, do you know why?
Whanted to setup the packagist hook.

**Comment by ElectricMaxxx** [comment]
Really? I created it. :-)

**Comment by wouterj** [comment]
Yeah, for some reason you have the permission to create repository, but you don't have global owner permissions. To be very honest, I don't really understand GitHub's organization permission system (and the Symfony CMF teams are a complete mess, we really need to clean this up one day).

**Comment by ElectricMaxxx** [comment]
Looks like we are really good on track now. Sonata gets its last clean ups. Most bundles needs its UPGRADE file only. 
Just create and routing-auto got some open issues or ???. So what to do with them?

**Comment by wouterj** [comment]
CreateBundle: It's updated to allow Sf 3 and FOSRestBundle 2. I think we should do the normal clean-ups here (directory, class parameters, etc.) and release 2.0. We can always do the feature enhancements in 2.x.

RoutingAutoBundle: It has some nice new features. Let's do the remaining clean-ups here and release 2.0 also.

Besides this, documentation is a major issue at the moment. We can't release 2.0 without first looking at the docs properly and revisiting everything that needs revisiting. Things that needs to be done in the docs for any bundle:

 1. Create SonataAdminIntegrationBundle docs
 2. Move all sonata admin related stuff to the new bundle's docs
 3. Quickly review all bundle docs and modify things where needed.
 4. Recheck the quick tour and tutorial, they are the starting points and need to be correct.

(1) and (2) have to be done before releasing an RC version.

Giving a christmas gift is nice, but I'm afraid releasing RC versions between christmas and new year isn't going to be very successful. I'm afraid not many people are working then, so the versions won't be tested that much on real apps.

So I would suggest the following planning:

Date | Deadline
--- | ---
**December 23th** | Feature freeze
**December 24th - December 30th** | Work on the docs (1 and 2) and tweaks/fixes (like directory structure changes, UPGRADE guides, etc); Write a blog post
**January 2th** | Release RC 1
**January 2th - January 6th** | Spread the news; test the RC; fix bugs; work on docs (3 and 4)
**January 6th** | Release stable (if no major bugs were found)

**Comment by ElectricMaxxx** [comment]
Thanks @wouterj looks like a good plan.

> Giving a christmas gift is nice, but I'm afraid releasing RC versions between christmas and new year isn't going to be very successful. I'm afraid not many people are working then, so the versions won't be tested that much on real apps.

:+1: Would have been really nice, but you are right from an testing perspective.

Can we do a nice spreadsheet for the docs too? Or just tagging issues with milestones? (currently i am a scrummaster in my team, so i really like to see stuff on a board.) Just for visualisation?

**Comment by dbu** [comment]
agree with that plan!

lets add the documentation upgrades to the spreadsheet i would suggest.

@ShinDarth maybe the CreateBundle or RoutingAutoBundle cleanups could be
something for you? one task is looking where we use <parameter>... in
the service configurations and remove them, and instead hardcode the
class names in the service configuration, as per symfony best practices.

**Comment by wouterj** [comment]
Here you go: https://github.com/symfony-cmf/symfony-cmf-docs/projects/1

**Comment by wouterj** [comment]
So I did some work on routing auto and create today. I think we're about ready now.

Shall we start releasing beta1 versions of bundles that are ready? (i.e. have all fields green in the spreadsheat and no important open issues)

**Comment by ElectricMaxxx** [comment]
Except from travis's build limit it would be easier when doing them all together from my POV. Doing so we will be able to put them all together into sandbox and standard-edition to see tests going green and anounce it by blog post.

**Comment by wouterj** [comment]
Created a GitHub organization-wide project to keep track of everything we need to do before releasing the first RC versions this week: https://github.com/orgs/symfony-cmf/projects/1

If something is missing from this project, please add it. Let's focus on getting everything in the ready column and getting things merged.

For now, enjoy your christmas days! 🎄

**Comment by ElectricMaxxx** [comment]
I did the same on waffle.io: https://waffle.io/symfony-cmf/symfony-cmf?milestone=2.0,2.2,1.0

**Comment by wouterj** [comment]
Something that just occured to me: Puli still isn't stable, so we can't release Resource packages stable, so we can't release TreeBrowserBundle stable and SonataDoctrinePhpcrOdmBundle stable.

So instead, we should stop focussing on the SonataAdminIntegrationBundle. We should focus on getting all other main bundles ready for release. This means only RoutingAuto has some stuff left that should be done before RC1.

**Comment by ElectricMaxxx** [comment]
But we can release sonata-admin-Integration without the new treebrowser?

Or is there something we can do to get the release of Puli done?

**Comment by ElectricMaxxx** [comment]
Adding some deeper information into the upgrade files and doing same on meun-bundle, let the question grow in my head: "Can we tag 2.1 for menu-bundle or schould be tag 3.0?". whatsching the changes in https://github.com/symfony-cmf/menu-bundle/pull/280 let me think, that those Changes are BC-Breaks.

**Comment by ElectricMaxxx** [comment]
@dbu  there are more open issues in components not in the cmf organisation (and not in the sheet of @wouterj). Can you go through them too (as you did where i asked directly). You can get an overview (mostly) on: https://waffle.io/symfony-cmf/symfony-cmf?milestone=2.0,2.2,1.0

**Comment by dbu** [comment]
i went over waffle and removed some from the milestones and commented on others. i think we should focus on the cmf bundles and leave sonata / admin stuff for later. 

regarding menu bundle: indeed, we seem to have released a 2.0 already ( i think to adjust to knp menu 2.0) so we should up the version to 3 now

**Comment by wouterj** [comment]
All CMF core bundles (not the sonata stuff) are ready for RC imo. Shall I tag them with RC1 tonight/tomorrow?

 > regarding menu bundle: indeed, we seem to have released a 2.0 already ( i think to adjust to knp menu 2.0) so we should up the version to 3 now

Do we really want 3.0-RC1 or stick with 2.1-RC1?

**Comment by ElectricMaxxx** [comment]
> Do we really want 3.0.0-RC1 or stick with 2.1.0-RC1?
:+1: for 3.0.0-RC1

**Comment by ElectricMaxxx** [comment]
Do not really know if relevant for RC, but we should also update the UPGRADE files of the bundles with the naming changes in sonata-integration.

**Comment by wouterj** [comment]
> Do not really know if relevant for RC, but we should also update the UPGRADE files of the bundles with the naming changes in sonata-integration.

Ah, thanks. Will check the files first in that case.

**Comment by wouterj** [comment]
Tagged the RC1 release for all components today, this means signed released for:

 * Slugifier API
 * Testing
 * Routing
 * RoutingAuto

For the bundles, we first have to get the CmfCoreBundle ready. This means finishing https://github.com/symfony-cmf/core-bundle/pull/220 and https://github.com/symfony-cmf/core-bundle/pull/214

**Comment by wouterj** [comment]
Imo, such changes can be done in the RC phase

**Comment by dbu** [comment]
working on core bundle, i realize that phpunit complains about `getMock`, except with php 5.5 that uses an older version of phpunit. which brought me to realize that maybe we could drop php 5.5 support. i know that symfony 3 supports 5.5.9 but as we do a new release now, i see no problem not supporting it. 5.5 [gets no more security updates](http://php.net/supported-versions.php) so its a severe risk to run it in production.

**Comment by ElectricMaxxx** [comment]
I would go one step further and drop 5.6 to as it isn't supported
anymore. But @WouterJ said something about LTS on our bundles/component,
right?

**Comment by dbu** [comment]
well, 5.6 still gets security updates and people might be stuck with it
for some reason (no control over the hosting platform, ...)

**Comment by wouterj** [comment]
+1 for dropping PHP 5.5 support and using `createMock()`.

**Comment by ElectricMaxxx** [comment]
:+1: for dropping 5.5 and for using `[]` only then :-)

**Comment by wouterj** [comment]
Doctrine PHPCR ODM 2.0 is very important for the release, I can't release any RC of the bundles without having at least 2.0.0-RC1. I tried to ping on a PR on the PHPCR ODM repo, but didn't get any response there.

@dbu @lsmith77 @dantleech what has to be done to get 2.0 to an RC state?

**Comment by ElectricMaxxx** [comment]
When the components have valid milestonses set, i see one [issue](https://waffle.io/symfony-cmf/symfony-cmf?source=doctrine%2Fphpcr-odm,doctrine%2Fphpcr-odm-documentation,doctrine%2FDoctrinePHPCRBundle&milestone=2.0) only for phpcr-odm, right @dbu?

**Comment by ElectricMaxxx** [comment]
@dbu we still have lots of [open issues with 2.0 milestone](https://waffle.io/symfony-cmf/symfony-cmf?milestone=2.0&source=sonata-project%2FSonataDoctrinePhpcrAdminBundle) in `SonataDoctrinePhpcrAdminBundle`? Are they relevant when releasing RC-1 there. I think we depend in sonata-phpc-admin-integration-bundle on that.

**Comment by dbu** [comment]
first off, admin should not prevent us from releasing cmf 2.0. the cmf
admin integration can come later. i would first focus on the cmf bundles
and tackle admin once we have RC1.

some of the issues in the waffle seem not necessarily 2.0 blockers to
me, could also go into 2.1 or whatever, if somebody ever finds the time
to work on them.

**Comment by dbu** [comment]
i am tracking the release process in #241

**Comment by shkkmo** [comment]
I may be late to the discussion regarding dropping PHP 5.5 support, but I thought I'd point out that Ubuntu 14.04 uses PHP 5.5 and doesn't support PHP 5.6 in the official repositories (nor does 16.04 which uses PHP 7). See: http://askubuntu.com/questions/527533/when-will-php-5-6-be-in-the-official-canonical-repos

**Comment by dbu** [comment]
php 5.5 gets no security fixes anymore since summer 2016. 5.6 is in
security fixes only mode. http://php.net/supported-versions.php

imho it is bad practice to keep supporting insecure versions of php.
there is still the old version for people stuck on legacy systems. but
the more pain it is to stay on legacy systems, the more incentive to
upgrade the base system...

from a maintainer point of view, php 5.6 has features that 5.5 is
missing. its easier to support fewer different versions.

**Comment by shkkmo** [comment]
Canonical will be backporting security fixes into PHP 5.5 for the life of Ubuntu 14.04 which is an LTS and supported through 2019, this means it is not  necessarily a bad practice to continue using it in production.

IMHO, it is bad practice to deviate from the base framework requirements unless there is a good reason why the new requirements are needed. Obviously I am not a maintainer so don't really get a vote on if the benefits of short array notation and createMock qualify as such a reason.

However, this requirement bump leaves us with the choice of going with a unofficial PPA for PHP 5.6 and losing many of the benefits of being on a LTS release, moving to a non LTS release such as 15.04, or waiting to upgrade many of our dependencies until they all support PHP 7 and we can move to 16.04

**Comment by dbu** [comment]
yeah, this is the quandary between stability and innovation. in my experience, the performance gains of php 7 are well worth it to do efforts to upgrade the code base to php 7. most of the symfony ecosystem does support php 7, for example.

i am sorry that we put you into a problem here. but indeed for us its too late to go back to support php 5.5, we want to release the work we did and have limited resources.
